pkg/cni/plugin.go

```diff
@@ -0,0 +1,259 @@
+type EnvArgs struct {
+	types.CommonArgs
+	MAC types.UnmarshallableString `json:"mac,omitempty"`
+}
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func loadConf(bytes []byte) (*NetConf, string, error) {
+	n := &NetConf{}
+	if err := json.Unmarshal(bytes, n); err != nil {
+		return nil, "", fmt.Errorf("failed to load netconf: %v", err)
+	}
+
+	return n, n.CNIVersion, nil
+}
+
+func getEnvArgs(envArgsString string) (EnvArgs, error) {
+	if envArgsString != "" {
+		e := EnvArgs{}
+		err := types.LoadArgs(envArgsString, &e)
+		if err != nil {
+			return EnvArgs{}, err
+		}
+		return e, nil
+	}
+	return EnvArgs{}, nil
+}
+
+func getMTUByName(ifName string) (int, error) {
+	link, err := netlink.LinkByName(ifName)
+	if err != nil {
+		return 0, err
+	}
+	return link.Attrs().MTU, nil
+}
+
+func configureArp(macvtapLink netlink.Link, netns ns.NetNS) error {
+	err := netns.Do(func(_ ns.NetNS) error {
+		// For sysctl, dots are replaced with forward slashes
+		ifaceNameAllowingDots := strings.Replace(macvtapLink.Attrs().Name, ".", "/", -1)
+
+		// TODO: duplicate following lines for ipv6 support, when it will be added in other places
+		ipv4SysctlValueName := fmt.Sprintf(IPv4InterfaceArpProxySysctlTemplate, ifaceNameAllowingDots)
+		if _, err := sysctl.Sysctl(ipv4SysctlValueName, "1"); err != nil {
+			// remove the newly added link and ignore errors, because we already are in a failed state
+			netlink.LinkDel(macvtapLink)
+			return fmt.Errorf("failed to set proxy_arp on newly added interface %q: %v", macvtapLink.Attrs().Name, err)
+		}
+		return nil
+	})
+	return err
+}
+
+func configureMacvtap(conf *NetConf, macAddr net.HardwareAddr, ifName string, netns ns.NetNS) (*current.Interface, error) {
+	macvtapIface, err := netlink.LinkByName(conf.DeviceID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup device %q: %v", conf.DeviceID, err)
+	}
+
+	// move the macvtap interface to the pod's netns
+	if err := netlink.LinkSetNsFd(macvtapIface, int(netns.Fd())); err != nil {
+		return nil, fmt.Errorf("failed to move iface %s to the netns %d because: %v", macvtapIface, netns.Fd(), err)
+	}
+
+	// configure the MTU of the macvtap iface
+	err = netns.Do(func(_ ns.NetNS) error {
+		if conf.MTU != 0 {
+			if err := netlink.LinkSetMTU(macvtapIface, conf.MTU); err != nil {
+				return fmt.Errorf("failed to set the macvtap MTU for %s: %v", conf.DeviceID, err)
+			}
+		}
+
+		if macAddr.String() != "" {
+			if err = netlink.LinkSetHardwareAddr(macvtapIface, macAddr); err != nil {
+				return fmt.Errorf("failed to add hardware addr to %q: %v", ifName, err)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	macvtap := &current.Interface{Name: ifName}
+	err = configureArp(macvtapIface, netns)
```

Comment:
The function is performing the whole time in the netns. Could you follow what you do in the previous block, stay in the netns and execute a simple configureArp(macvtapIface)?

Reply:
Done.